recipes/src/recipes/RecipeComponent.js

```diff
@@ -0,0 +1,94 @@
+import React,  { useState, useEffect, } from "react";
+
+import styles from '../styles';
+import RecipeList from './RecipeList';
+import RecipeDetail from './RecipeDetail';
+import RecipeForm from './RecipeForm';
+
+import recipesApi from "./recipesApi";
+
+
+const {
+  Container,
+  AddButton,
+} = styles;
+
+const {
+  getRecipes: getRecipesFromDb,
+  deleteRecipe: deleteRecipeFromDb,
+  addOrUpdateRecipe: addOrUpdateRecipeInDb,
+} = recipesApi;
+
+
+function RecipeComponent() {
+
+  const [recipes, setRecipes] = useState([]);
+  const [selected, setSelected] = useState(false);
+  const [updating, setUpdating] = useState(false);
+  const [refreshing, setRefreshing] = useState(false);
+
+  const recipeDbUpdated = () => {
+    setUpdating(false);
+    setRefreshing(true);
+  }
+
+  async function deleteRecipe(id) {
+    await deleteRecipeFromDb(id);
+    recipeDbUpdated();
+  }
+
+  async function addOrUpdateRecipe(recipe) {
+    await addOrUpdateRecipeInDb(recipe);
+    recipeDbUpdated();
+  }
+
+  useEffect(() => {
+    let cancelled = false;
+
+    const getRecipes = async () => {
+      try {
+        const response = await getRecipesFromDb();
+
+        if(!cancelled) {
+          setSelected(false);
+          setRecipes(response.data);
+        }
+      } catch(e) {
+        if(!cancelled) {
+          // TODO Handle error state
+          alert(`GET recipes error: ${JSON.stringify(e)} ${
+            cancelled? 'after cancellation':''
+          }`);
+        }
+      }
+
+      if(!cancelled) {
+        setUpdating(false);
+        setRefreshing(false);
+      }
+    }
+
+    getRecipes();
+
+    return (() => {cancelled = true;});
```

Comment:
That looks a lot clearer/straightforward.
I got that from some blog - the idea, if I understood it, is that if the effect is called a second time before the first completes it doesn't try to un-set the state. It was associated with some error handling in the article I got it from.
 
I thought that the function passed to useEffect() had to return a function or nothing - making it async would return a promise, and it would have to be async to await the getRecipes call. 

I may be working on old information though - it makes things much simpler if it can just be passed an async function.

Reply:
Yep - that looks to work fine, so I guess I was working on old or spurious information there 😄